Add annotation field ``requiresVirtualLookup``

Marenz opened the pull request:
Not very confident in my code here, please check carefully.

Marenz committed:
Add annotation field ``requiresVirtualLookup``

chriseth commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1460,6 +1465,7 @@ bool ExpressionCompiler::visit(MemberAccess const& _memberAccess)
 		if (type.isSuper())
 		{
 			solAssert(!!_memberAccess.annotation().referencedDeclaration, "Referenced declaration not resolved.");
+			solAssert(_memberAccess.annotation().requiresVirtualLookup, "");
```
Can you add another field (or turn this into an enum) for "superLookup"?

chriseth commented on libsolidity/ast/ASTAnnotations.h:
```diff
@@ -236,6 +236,8 @@ struct UserDefinedTypeNameAnnotation: TypeNameAnnotation
 	/// Referenced declaration, set during reference resolution stage.
 	Declaration const* referencedDeclaration = nullptr;
+	/// Whether a virtual lookup needs to be performed to find the declaration.
+	bool requiresVirtualLookup = false;
```
Can this one ever be true?

Marenz replied:
Good point. I worked backwards and looked where we do a `resolveVirtual` in the codegen and it seems that method didn't lead to a case where a `UserDefinedTypeName` was resolved..

Marenz replied:
In fact, there is no mentioning of `UserDefinedTypeName` anywhere in `codegen/*`

chriseth replied:
Yes, I think this is just a "type checking" concept. Codegen only uses `annotation().type`

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -2656,6 +2656,9 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
 				"Using \"." + memberName + "(...)\" is deprecated. Use \"{" + memberName + ": ...}\" instead."
 			);
+
+		if (!funType->bound())
```
And if it is bound, the field is not set at all?

Marenz replied:
Well, then it's the default, which is `false`. I could be more explicit I suppose and set it if bound to `false`

Marenz replied:
The `optional`/`SetOnce` PR is a different one ;)

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -2656,6 +2656,9 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
 			);
+
+		if (!funType->bound())
+			annotation.requiresVirtualLookup = !!dynamic_cast<ContractType const*>(exprType);
```
Can you name examples for both cases?

Marenz replied:
As far as I understand:
```
contract A {
    function f() virtual public {}
}
contract B is A {
    function f() virtual override public {}
    function x() public { 
        A.f(); // This would be a `false` case because the type would be `TypeType` 
        this.f(); // This would be a `true` case because `this` should be the `ContractType` 
}
```

chriseth replied:
Actually `this.f()` is an external call and we should not do virtual lookup for external calls - so both of them should be non-virtual.

ekpyron replied:
I guess valid lookups in ``this`` will have non-empty ``possibleMembers`` and not end up here at all? But yeah, it's a bit non-transparent so far :-)...

Marenz replied:
I think that's also my problem, I don't know all those cases. I basically derived this behavior based on the backend `resolveVirtual` and `superFunction` calls

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -2523,8 +2523,10 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
 	TypePointer exprType = type(_memberAccess.expression());
 	ASTString const& memberName = _memberAccess.memberName();
 
+	auto& annotation = _memberAccess.annotation();
```
I think it's butter to call the function twice instead of adding another variable.

Marenz replied:
that may be, but I didn't add another one, this is just moved up from line 2543

Marenz replied:
in this case to take advantage of the existing variable even earlier. ( line 2529)

chriseth replied:
Ah right! I still think it's unnecessary, but you can also keep it.

chriseth commented on libsolidity/ast/ASTEnums.h:
```diff
@@ -30,6 +30,9 @@
 {
 
+/// Possible lookups for function resolving
+enum class VirtualLookup { None, Virtual, Super };
```
```suggestion
enum class VirtualLookup { Static, Virtual, Super };
```

chriseth commented on libsolidity/ast/Types.h:
```diff
@@ -1340,7 +1340,9 @@ class FunctionType: public Type
 	bool const m_valueSet = false; ///< true iff the value to be sent is on the stack
-	bool const m_bound = false; ///< true iff the function is called as arg1.fun(arg2, ..., argn)
+	/// true iff the function is called as arg1.fun(arg2, ..., argn)
+	/// e.g. false for contract functions or free functions.
```
Actually that's not necessarily true - it depends on how we will implement `using` in the future. So maybe to clarify that this is about the type of a variable / expression and not about the function declaration itself.

Marenz replied:
Hmm I am not sure how I would phrase that. Could you elaborate your example?

chriseth replied:
true iff the function is called as arg1.fun(arg2, ..., argn). This is achieved through the "using for" directive.

chriseth commented:
Needs rebase.

Marenz commented:
Update:
* Rebased
* Made new field also `SetOnce`
* Renamed Enum field `None` to `Static`

still missing:
* properly setting (and using) the field for UserDefinedTypeNameAnnotation
* proper rephrasing for m_bound

Marenz commented:
This is now ready
